src/full/decoder.rs

```diff
@@ -138,39 +138,66 @@ impl Decoder {
+    ///
+    /// Use this over `get_decoded_data` when you want to avoid heap allocation
+    /// and have a pre-allocated buffer ready.
+    pub fn get_decoded_data_into(self, out_buf: &mut [u8]) -> Result<usize, RLNCError> {
         if !self.is_already_decoded() {
             return Err(RLNCError::NotAllPiecesReceivedYet);
         }
 
-        let full_coded_piece_len = self.required_piece_count + self.piece_byte_len;
-        let mut decoded_data = Vec::with_capacity(self.piece_byte_len * self.required_piece_count);
-
-        // Iterate over the useful rows (which should be the decoded original pieces)
-        // and extract the data part from each row.
-        self.matrix.extract_data().chunks_exact(full_coded_piece_len).for_each(|full_decoded_piece| {
-            // The data part of the row starts after the coefficient columns.
-            let decoded_piece = &full_decoded_piece[self.required_piece_count..];
-            decoded_data.extend_from_slice(decoded_piece);
-        });
-
-        // Find the boundary marker to trim padding.
-        let last_index_of_decoded_data = decoded_data.len() - 1;
-        let boundary_marker_rev_index = decoded_data
-            .iter()
-            .rev()
-            .position(|&byte| byte == BOUNDARY_MARKER)
-            .unwrap_or(last_index_of_decoded_data);
-        let boundary_marker_index = last_index_of_decoded_data - boundary_marker_rev_index;
-
-        if boundary_marker_index == 0 {
+        let required_len = self.piece_byte_len * self.required_piece_count;
+        if out_buf.len() < required_len {
+            return Err(RLNCError::InvalidOutputBuffer);
+        }
+
+        let mut current_pos = 0;
+        let full_coded_piece_len = self.get_full_coded_piece_byte_len();
+
+        // Write the decoded data piece by piece into the output buffer
+        for chunk in self.matrix.extract_data().chunks_exact(full_coded_piece_len) {
+            let payload = &chunk[self.required_piece_count..];
+            let end_pos = current_pos + self.piece_byte_len;
+            out_buf[current_pos..end_pos].copy_from_slice(payload);
+            current_pos = end_pos;
+        }
+
+        let final_len = Self::get_final_data_len(&out_buf[..current_pos])?;
+
+        // Instead of returning a Vec, we return the length of the valid data.
+        Ok(final_len)
+    }
+
+    /// Helper to find the boundary marker, validate padding,
+    /// and return the final length of the original data.
+    fn get_final_data_len(padded_data: &[u8]) -> Result<usize, RLNCError> {
+        let last_index = padded_data.len().saturating_sub(1);
+
+        let boundary_marker_rev_index = padded_data.iter().rev().position(|&byte| byte == BOUNDARY_MARKER).unwrap_or(last_index);
+
+        let boundary_marker_index = last_index - boundary_marker_rev_index;
+
+        // Perform the validation checks
+        if boundary_marker_index == 0 && !padded_data.is_empty() {
```

Comment:
I don't understand why we need to check for `!padded_data.is_empty()`. When might it trigger ?